docs/internal/abi.md

```diff
@@ -0,0 +1,444 @@
+
+Basically, if the exponent is **not** _all ones_, then the value represented is a 64-bit floating pointer number.
+If it is, then we use some bits to encode which type it is (the discriminator), and the rest of the bits left to encode the value.
+
+We use the sign bit to represent pointers, leaving the remaining 52-bits (the entire mantissa) to represent the address of the pointer.
+Most 64-bit architectures use 52-bits or less to encode addresses, so that's enough for our purposes.
+When the sign bit is zero, we use 4 high bits in the mantissa to encode all representable types, followed by a padding of `(52-4-sizeof(type))` bits, and the value taking the `sizeof(type)` lower bits.
+
+We can encode all 32-bit or less types as well as pointers and `F64`, but not 64-bit integers or higher.
+For those, we use the wide packing below.
+
+The list of the types that can be NaN-packed:
+* Pointer
+* F64
+* F32
+* U32
+* U16
+* U8
+* I32
+* I16
+* I8
+* Bool
+* ISize (on a 32 bit platform)
+* USize (on a 32 bit platform)
+
+In the following ways:
+```
+[*][***********][52 bit mantissa] = F64
+[1][11111111111][52 data bits] = Pointer
+[0][11111111111][0000][47*][1 data bit] = Bool
+[0][11111111111][0001][16*][32 data bits] = F32
+[0][11111111111][0010][40*][8 data bits] = I8
+[0][11111111111][0011][32*][16 data bits] = I16
+[0][11111111111][0100][16*][32 data bits] = I32
+[0][11111111111][0101][16*][32 data bits] = ISize (32)
+[0][11111111111][0110][40*][8 data bits] = U8
+[0][11111111111][0111][32*][16 data bits] = U16
+[0][11111111111][1000][16*][32 data bits] = U32
+[0][11111111111][1001][16*][32 data bits] = USize (32)
+```
+
+The order of types and their bit patterns are arbitrary.
+
+The remaining types that need to be wide-packed are:
+* I64
+* U64
+* I128
+* U128
+* ISize (on a 64 bit platform)
+* USize (on a 64 bit platform)
+
+Because we have more than 16 types in total, we need to add 1 extra bit to the discriminator for wide packing.
+We also need to add another bit for identifying pointers, which we can use the top bit like NaN-boxing.
+
+```
+------------------------- Types also represented in NaN-boxing
+[1][**][0][0000] Pointer
+[0][**][0][0000] Bool
+[0][**][0][0001] F32
+[0][**][0][****] ... // Same bit-pattern as above
+------------------------- New types, only in wide-packing
+[0][**][1][0000] I64
+[0][**][1][0001] U64
+[0][**][1][0010] I128
+[0][**][1][0011] U128
+[0][**][1][0100] ISize (on a 64 bit platform)
+[0][**][1][0101] USize (on a 64 bit platform)
+```
+
+This bit pattern is also arbitrary and can change in the actual implementation.
+
+Note that, in 64-bit architectures, pointers (and `F64`) have the same width on both native and NaN-boxed representations.
+All other types increase the width to at least 64-bit, even if they were `U8` to begin with.
+
+Also note that the pattern above is an extra byte, so all values will have the whole 64/128 bits of the payload to be encoded.
+
```

Comment:
When resorting to wide-boxing in the heap, we presumably pad to alignment (which, on CHERI, means unions are really uncomfortably big, with a 16-byte header... OTOH, for some small solace, there's more room for cap-packing non-pointer types)?  When upcasting to a union type as part of a function call, do we follow SROA and pass the discriminator and payload in potentially different registers if they fit?

Reply:
> When resorting to wide-boxing in the heap, we presumably pad to alignment (which, on CHERI, means unions are really uncomfortably big, with a 16-byte header...

Correct.

> OTOH, for some small solace, there's more room for cap-packing non-pointer types)? When upcasting to a union type as part of a function call, do we follow SROA and pass the discriminator and payload in potentially different registers if they fit?

There are some discussions on SROA and alternatives at the end of the document, clearly inspired by this glaring issue. :)

Reply:
I think the key point is that, when we're doing wide packing, the fact that we don't have concurrent mutation means that there's no need to keep the data and the discriminator together.  When doing class / array layout, we can store them separately and pack the discriminator just as we would any U16 field.

Reply:
Which is explained below, in this document.